### PR TITLE
Fix items not filtered by Worksheet Template's method in Add analyses

### PR DESCRIPTION
**Pull request by Bugerman58**
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/1989

## Current behavior before PR

1. Filter does not apply
2. Method not showing in the view on worksheet template page

## Desired behavior after PR is merged

1. Filtering analyses by method from the worksheet 
2. Showing method in the view on worksheet template page

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/browser/worksheet/tools.py [hunk 1]
```diff
@@ -93,3 +95,15 @@ def getWorksheetLayouts():
         map(lambda entry: layouts.append(entry), layout_utility.getLayouts())
 
     return DisplayList(tuple(layouts))
+
+
+def getServiceUidsByMethod(method):
+    query = {
+        "portal_type": "AnalysisService",
+        "ia_active": True,
```
Comment (xispa) [inline src/bika/lims/browser/worksheet/tools.py]: typo: `ia_active` --> `is_active`

File: CHANGES.rst [hunk 2]
```diff
@@ -4,6 +4,7 @@ Changelog
 2.2.0 (unreleased)
 ------------------
 
+- #1990 Issue creating worksheet with method specific template
```
Comment (xispa) [inline CHANGES.rst]: Please be more specific. Something like "Fix items not filtered by Worksheet Template's method in Add analyses"

File: src/bika/lims/browser/worksheet/views/add_analyses.py [hunk 3]
```diff
@@ -114,6 +115,12 @@ def __init__(self, context, request):
 
     def __call__(self):
         super(AddAnalysesView, self).__call__()
+        wst = self.context.getWorksheetTemplate()
+        if wst:
+            method = wst.getRawRestrictToMethod()
+            # restrict the available analysis services by method
+            if method:
+                self.contentFilter["getServiceUID"] = getServiceUidsByMethod(method)
```
Comment (xispa) [inline src/bika/lims/browser/worksheet/views/add_analyses.py]: Better move all this inside the function `update` so the filter applies from the start. Note the `update` function is called by the superclass `ListingView` in `__call__`: https://github.com/senaite/senaite.app.listing/blob/2.x/src/senaite/app/listing/view.py#L220

**Review by xispa — CHANGES_REQUESTED** [review]
Thanks @Bugerman58 , some changes requested!

File: src/bika/lims/content/worksheettemplate.py [hunk 4]
```diff
@@ -100,7 +99,7 @@
         allowed_types=("Method",),
         relationship="WorksheetTemplateMethod",
         referenceClass=HoldingReference,
-        widget=SelectionWidget(
```
Comment (xispa) [inline src/bika/lims/content/worksheettemplate.py]: Why? This does not seem related to the issue you want to address. See comment I've added a few minutes ago to this other PR from @Lunga001 : https://github.com/senaite/senaite.core/pull/1887/files#r861569889 . Note that if you change this widget, the filtering of "Instruments" won't work

File: src/bika/lims/browser/worksheet/views/add_analyses.py (hunk 3 above)
Comment (xispa) [reply src/bika/lims/browser/worksheet/views/add_analyses.py]: Also take into account that `getServiceUidsByMethod` may return an empty list

File: src/bika/lims/browser/worksheet/tools.py (hunk 1 above)
Comment (Bugerman58) [reply src/bika/lims/browser/worksheet/tools.py]: oh, sorry )

File: CHANGES.rst (hunk 2 above)
Comment (Bugerman58) [reply CHANGES.rst]: done

File: src/bika/lims/content/worksheettemplate.py (hunk 4 above)
Comment (Bugerman58) [reply src/bika/lims/content/worksheettemplate.py]: ok. i will undo the changes
done

File: src/bika/lims/browser/worksheet/views/add_analyses.py (hunk 3 above)
Comment (Bugerman58) [reply src/bika/lims/browser/worksheet/views/add_analyses.py]: is it right, because if method not empty in the worksheet and not exists samples with this method than nothing show(empty list)
or i wrong?

File: src/bika/lims/browser/worksheet/tools.py [hunk 5]
```diff
@@ -93,3 +95,15 @@ def getWorksheetLayouts():
...
+def getServiceUidsByMethod(method):
+    query = {
+        "portal_type": "AnalysisService",
+        "is_active": True,
+        "method_available_uid": method,
+    }
+    setup_catalog = api.get_tool(SETUP_CATALOG)
+    uids = map(lambda s: s.UID, setup_catalog(query))
+
+    return uids
```
Comment (xispa) [inline src/bika/lims/browser/worksheet/tools.py]: Please move this function to `add_analyses.py`. Also note you assume the value for `method` is a always UID. Worth to check and resolve the value for `method` first.

**Review by xispa — COMMENTED** [review]
thanks @Bugerman58 , you are almost there!

File: src/bika/lims/browser/worksheet/tools.py (hunk 5 above)
Comment (Bugerman58) [reply src/bika/lims/browser/worksheet/tools.py]: done

File: src/bika/lims/browser/worksheet/views/add_analyses.py [hunk 6]
```diff
@@ -32,11 +32,27 @@
...
+    if not api.is_uid(method):
+        method = api.get_uid(method)
+
+    query = {
+        "portal_type": "AnalysisService",
+        "is_active": True,
+        "method_available_uid": method,
+    }
+    setup_catalog = api.get_tool(SETUP_CATALOG)
+    uids = map(lambda s: s.UID, setup_catalog(query))
```
Comment (ramonski) [inline src/bika/lims/browser/worksheet/views/add_analyses.py]: better use this:
`uids = map(api.get_uid, setup_catalog(query)`

**Review by ramonski — CHANGES_REQUESTED** [review]

File: src/bika/lims/browser/worksheet/views/add_analyses.py (hunk 6 above)
Comment (Bugerman58) [reply src/bika/lims/browser/worksheet/views/add_analyses.py]: done
